common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java

```diff
@@ -231,8 +231,22 @@ public UrlArgument workingRepositoryUrl() {
         return new UrlArgument(runOrBomb(gitConfig).outputForDisplay().get(0));
     }
 
-    public static void checkConnection(UrlArgument repoUrl) {
-        commandToCheckConnection(repoUrl).runOrBomb(repoUrl.forDisplay());
+    public static void checkConnection(UrlArgument repoUrl, String branch) {
+        CommandLine commandLine = git().withArgs("ls-remote").withArg(repoUrl);
+        ConsoleResult result = commandLine.runOrBomb(repoUrl.forDisplay());
+        if(!isValidBranch(result.outputAsString(), branch)){
+            throw invalidBranchException(repoUrl.forDisplay(), branch);
```

Comment:
You seem to be checking if the output contains a string vs testing if the output contains a pattern `refs/heads/FOO$`

Your code will likely match a substring. So a branch `mas` will match master for e.g.